google/cloud/forseti/scanner/audit/retention_rules_engine.py

```diff
@@ -0,0 +1,388 @@
+
+    def generate_bucket_violation(self, bucket):
+        """generate a violation.
+
+        Args:
+            bucket (Bucket): The bucket that triggers the violation
+        Returns:
+            RuleViolation: The violation
+        """
+        data_lifecycle = bucket.get_lifecycle_rule()
+        data_lifecycle_str = json.dumps(data_lifecycle)
+
+        return self.RuleViolation(
+            resource_name=bucket.id,
+            resource_type=bucket.type,
+            full_name=bucket.full_name,
+            rule_name=self.rule_name,
+            rule_index=self.rule_index,
+            violation_type=VIOLATION_TYPE,
+            violation_data=data_lifecycle_str,
+            resource_data=bucket.data,
+        )
+
+    def find_violations(self, res):
+        """Get a generator for violations
+        Args:
+            res (Resource): A class derived from Resource
+        Returns:
+            Generator: All violations of the resource breaking the rule.
+        """
+
+        if res.type == 'bucket':
+            return self.find_violations_in_bucket(res)
+        return iter(())
+
+    def bucket_max_retention_violation(self, bucket):
+        """Get a generator for violations especially for maximum retention
+        Args:
+            bucket (bucket): Find violation from the bucket
+        Yields:
+            RuleViolation: All max violations of the bucket breaking the rule.
+        """
+        if self.max_retention is None:
+            return
+
+        # There should be a condition which guarantees to delete data
+        exist_max_limit = False
+        bucket_lifecycle = bucket.get_lifecycle_rule()
+        if bucket_lifecycle is None:
+            yield self.generate_bucket_violation(bucket)
+        else:
+            for lc_item in bucket_lifecycle:
+                if lc_item.get('action', {}).get('type') == 'Delete':
+                    conditions = lc_item.get('condition', {})
+                    age = conditions.get('age')
+                    if age is not None and len(conditions) == 1:
+                        # the config does not have conditions other than age
+                        if age <= self.max_retention:
+                            exist_max_limit = True
+                            break
+            if not exist_max_limit:
+                yield self.generate_bucket_violation(bucket)
+
+    def bucket_min_retention_violation(self, bucket):
+        """Get a generator for violations especially for minimum retention
+        Args:
+            bucket (bucket): Find violation from the bucket
+        Yields:
+            RuleViolation: All min violations of the bucket breaking the rule.
+        """
+        if self.min_retention is None:
+            return
+
+        bucket_lifecycle = bucket.get_lifecycle_rule()
+        if bucket_lifecycle is not None:
+            for lc_item in bucket_lifecycle:
+                if lc_item.get('action', {}).get('type') != 'Delete':
```

Comment:
If you move the check on `age` into `bucket_conditions_guarantee_min`, then the body of this loop becomes:
```
if (lc_item.get('action', {}).get('type') == 'Delete' and
    not bucket_conditions_guarantee_min(lc_item.get('condition'), self.min_retention):
  yield self.generate_bucket_violation(bucket)
```